### ISSUE TITLE
SVG image support for HTML

### ISSUE THREAD

**Issue by ruslo** [sphinx-doc/sphinx#2240]
Currently if I use `.. image:: myfile.svg` I've got image as a whole and can't click on elements of SVG picture. I can open image separately and then it works. It would be nice to have SVG with clickable elements.

**Comment by tk0miya** [comment]
In my short investigation:

Now, Sphinx uses `<img>` tag for embed SVG images.
It disables hyperlinks in SVG images.
To resolve this, use `<object>` tag instead.
But it causes new issues.
1. Sphinx generates hyperlink to original image like `<a><img /></a>` if image is resized. but after replacing with `<object>`, it does not work correctly.
2. In `.svg` file, The `target` parameter is required at all hyperlinks. (if none, hyperlinks do not work)

And more, I worry about resizing parameter (width and height) for `<object>` tag. I don't know they are always work correctly in any browsers, for any SVG images, and so on (of course, simple case looks fine for me :-)

Please give me some advice.

**Comment by ruslo** [comment]
> Please give me some advice

Sorry, HTML stuff not my specialization, have very poor knowledge about peculiarities.
I've asked SO question, it has a few links that may be helpful:
- http://stackoverflow.com/questions/34777943

**Comment by mb21** [comment]
I don't think there's a lot that Sphinx can do here—it's really [quite complicated](http://stackoverflow.com/questions/11374059/make-an-html-svg-object-also-a-clickable-link). Although you could consider adding an option to the `.. image` directive that lets the user specify whether to render as an `img` or `object` tag.

In the meantime, people can write raw HTML as a workaround I guess:

```
.. raw:: html

    <object data="myfile.svg" type="image/svg+xml"></object>
```

**Comment by ruslo** [comment]
> people can write raw HTML as a workaround I guess

Thanks for suggestion, I've created simple project for testing:
- https://github.com/forexample/test-sphinx-svg-image

problem I've found so far is that when you resize windows browser there is no slider appear, so some svg nodes are not accessible:

![screen](https://cloud.githubusercontent.com/assets/4346993/13233790/06190136-d9e8-11e5-8a1e-3fc6c5e6009d.png)

Is there any simple solutions/workaround for this?

**Comment by mb21** [comment]
@ruslo, the following CSS should work:

```
.section {
  overflow: scroll;
}
```

**Comment by ruslo** [comment]
> the following CSS should work

@mb21 Have no idea what does it mean and how it can be applied, can you explain patch in terms of RST code, thanks and sorry :)

**Comment by ruslo** [comment]
Here is page with SVG image that works well for me:
- http://en.cppreference.com/w/cpp/io

As far as I understand it uses `<img src="*.svg"></img>`

**Comment by mb21** [comment]
Well, they use an `img` and a `<map>`... instead of clickable svg.

About adding CSS: put the above code snipped in `_static/style.css` and follow http://stackoverflow.com/questions/23462494/how-to-add-custom-css-file

**Comment by ruslo** [comment]
> put the above code snipped in _static/style.css and follow

@mb21 Here is the change: https://github.com/forexample/test-sphinx-svg-image/commit/65674884d06c893c46b53aaa0e35aee594e600c5

Result of the rendering on local machine:

![screen](https://cloud.githubusercontent.com/assets/4346993/13246945/5892e25e-da49-11e5-9e4b-0fd4a1ff00f9.png)

See no effects on rtfd.org: http://test-sphinx-svg-image.rtfd.org/en/latest/clickable-image.html

**Comment by mb21** [comment]
yeah, doesn't seem like the CSS was included... not sure why (I'm no sphinx expert :P)

**Comment by ruslo** [comment]
> yeah, doesn't seem like the CSS was included... not sure why (I'm no sphinx expert :P)

@mb21 Well to be honest even if it will be included somehow I will not say I really enjoy how it's rendered (see screenshot from my local machine).

**Comment by SimonBiggs** [comment]
To include clickable svg links within sphinx I did the following:

```rst
.. raw:: html
    :file: ../graphs/pymedphys_analysis.gamma.svg
```

See:

https://raw.githubusercontent.com/pymedphys/pymedphys/1915b9496e93782bdac7dcebff7e26e470e5ff57/docs/graphs/graphs.rst

This then freed me to write the following within an imported style sheet:

```css
svg {
    width: 100%;
}
```

https://github.com/pymedphys/pymedphys/blob/f4d404fa1cf3f551c4aa80ef27438f418c61a436/docs/_static/style.css

This made the svg fit the container as desired.

See:

https://pymedphys.com/developer/dependencies.html#pymedphys

If it helps, these additions were within the following pull request to PyMedPhys:

https://github.com/pymedphys/pymedphys/pull/285

**Comment by 2bndy5** [comment]
This may be off topic, but it seems to me that graphviz would help in @ruslo use case. I think each "node" in the graph could be hyperlinked...

**Comment by ain-soph** [comment]
Just curious, why not consider this feature in the upstream `docutils` rather than `sphinx` ? It'll be nice to have the new options in `image` directives.  
It seems they have been working on embedding images, and embedding svgs in the latest releases.  

But I've proposed an feature request about adding `object` as an option, which is refused.  
https://sourceforge.net/p/docutils/feature-requests/83/

**Comment by kpinc** [comment]
I filed a feature request (#99) with docutils.  See: https://sourceforge.net/p/docutils/feature-requests/99/

(I find it particularly annoying there is no work-around for figures.)

**Comment by kpinc** [comment]
The feature request with docutils is something of a mash-up of support for SVG functionality.  We'll see if it remains open, as the title of the feature request focuses on SVG scaling.

For the record, here is an example of Sphinx configuration and use which produces SVG images that scale.

----

An example of how to have SVG scale dynamically to browser
window width in a document generated by Sphinx v7.1.0 with the
default "alabaster" theme:

RST requires the "image" (or "figure") directive have the option:

:width: 100 %

The Sphinx "basic" theme, the parent class of all other themes,
requires a Sphinx html_theme_option key/value pair of:

'body_max_width': 'none'

The Sphinx "alabaster" theme, the default theme, requires a
Sphinx html_theme_option key/value pair of:

'page_width': '100%'

In a Sphinx config.py file you would see something like:

```
html_theme_options = {
    'body_max_width': 'none',
    'page_width': '100%',
}
```
Although you'd expect that the html_theme_options dict might
have other key/value pairs as well.

**Comment by kpinc** [comment]
Requested SVG hyperlink and event support in docutils feature request #100.
https://sourceforge.net/p/docutils/feature-requests/100/

**Comment by gmilde** [comment]
Direct embedding of SVG is now implemented in Docutils (repository version)
https://sourceforge.net/p/docutils/code/9517/  and Docutils 0.21 (2024-04-09).

There seem to be problems with image  `:loading: embed`  in Sphinx, though. See #12345.
